design/node-agent-concurrency.md

```diff
@@ -0,0 +1,78 @@
+
+## Non-Goals
+- VGDP instances from different nodes always run in concurrent since in most common cases the resources are isolated. For special cases that some resources are shared across nodes, there is no support at present
+- In practice, restores run in prioritized scenarios, e.g., disarster recovery. However, the current design doesn't consider this difference, a VGDP instance for a restore is blocked if it reaches to the limit of the concurrency, even though the ones block it are for backups. If users do meet some problems here, they should consider to stop the backups first
+- Sometimes, users wants to block VGDP instances from running in a specific node, this is not considered in the current design. To archive this, more modules need to be considered, just setting VGDP instances' concurrent number to 0 won't bring the ideal reasult
+
+## Solution
+
+### Global concurrent number
+We add a node-agent server parameter ```data-path-concurrent-num``` to accept a concurrent number that will be applied to all nodes if the per-node number is not specified.  
+The number starts from 1 which means there is no concurrency, only one instance of VGDP is allowed. There is no roof limit.    
+If the node-agent server parameter is not specified, a hard-coded default value will be used, the value is set to 1.   
+
+### Per-node concurrent number
+We allow users to specify different concurrent number per node, for example, users can allow 3 concurrent instances in Node-1, 2 instances in Node-2 and 1 instance in Node-3. This is for below considerations:
+- The resources may be different among nodes. Then users could specify smaller concurrent number for nodes with less resources while larger number for the ones with more resources
+- Help users to isolate critical environments. Users may run some critical workloads in some specified nodes, since VGDP instances may take large resource consumption, users may only want to run them in the other nodes or run less number of instances in the nodes with critical workloads
+
+The range of Per-node concurrent number is the same with Global concurrent number.  
+Per-node concurrent number is preferable to Global concurrent number, so it will overwrite the Global concurrent number for that node.  
+Per-node concurrent number is implemented in a separate configMap named ```node-agent-configs```. This configMap is not created by Velero, users should create it manually on demand. The configMap should be created before node-agent server starts, otherwise, node-agent server should be restarted in order to make the effect.  
+
+The ```node-agent-configs``` configMap may be used for other purpose of configuring node-agent in future, at present, there is only one kind of data ```data-path-concurrency```. Its data struct is like below: 
+
+```go
+type DataPathConcurrency struct {
+	// ConfigRules specifies the configs by node identified by label selector
+	ConfigRules map[string]int `json:"configRules"`
```

Comment:
Instead of a string, we should use a [LabelSelector](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector), IMO. 

Also, something to consider: if multiple labels are matched, which one is selected? I suggest the safe thing is the lesser of the entries.

Reply:
>  if multiple labels are matched, which one is selected? I suggest the safe thing is the lesser of the entries

Yes, that is what we are going to do, let me clarify in the design

Reply:
> Instead of a string, we should use a [LabelSelector](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector), IMO.

See [this comment](https://github.com/vmware-tanzu/velero/pull/6950#discussion_r1372505929). The string is in the pattern of LabelSelector, so the code will convert a string to a LabelSelector. This is for two considerations:
1. LabelSelector cannot be a key of map. To use LabelSelector, the structure will be complex
2. A string is convenient for users to write the config, they don't need to write "{}" or the assignment to "matchLabel"

Reply:
I don' know if this makes sense. There is a well-defined struct for label selector, and if you don't use that, it won't be very clear. 

I would suggest using a CRD for this and the well-known label selector struct. I would probably just do a simple list. Then, you can order that list by the size of the threads. you will loop through the selectors, and once one is found, it will be the lowest option. 

I don't think you gain anything from having a map here, as to see if a node is selected you have to look at the key, so you will have to loop over the map. I think that a list would be a better choice.

Reply:
Let me try with a list of label selector. I have ever considered this, but it eventually involves two data structures, so I just dropped

Reply:
Done, changed to LabelSelector